### PR TITLE
[SDK-1920] Add Calling an API chapter to Angular quickstart

### PR DESCRIPTION
**Pull request by stevehobbsdev**
* Calling an API chapter added
* Updates 01-Login to use variables in snippets (workshop feedback)
* Updates 01-Login auth button snippet to use ngIf/Else construct (workshop feedback)

### CODE REVIEW

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
+---
+title: Call an API
+description: This tutorial demonstrates how to make API calls for protected resources on your server.
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: the tutorial appears to show how to call our management API rather than on your server...

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+contentType: tutorial
+useCase: quickstart
+---
+<!-- markdownlint-disable MD002 MD034 MD041 -->
+
+When you use Auth0, you delegate the authentication process to a centralized service. Auth0 provides you with functionality to log in and log out users from your Angular application. However, your application may need to access resources from a secured API.
+
+You can also [protect an API with Auth0](https://auth0.com/docs/microsites/protect-api/protect-api). We offer multiple [Auth0 API quickstarts](https://auth0.com/docs/quickstart/backend) to help you integrate Auth0 with your backend platform.
+
+When you use Auth0 to protect your API, you delegate the authentication process to a centralized service that ensures only approved applications can access private resources. Your Angular application authenticates the user and receives an [access token from Auth0](https://auth0.com/docs/tokens/concepts/access-tokens). The application can then pass that access token to your API as a credential. In turn, your API can use [Auth0 libraries](https://auth0.com/docs/libraries) to verify the access tokens it receives from the calling application and issue a response with the desired data.
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: I would suggest rephrasing so that the description applies to any API protected by Auth0, as opposed to the current formulation that refers only to the custom API. Example (also note small tweaks like authn->authz):

When an API is protected by Auth0, the authorization process is delegated to a centralized service that ensures only approved applications can access private resources. Your Angular application authenticates the user and, in the process, it also receives an access token from Auth0. The application can then pass that access token to the API as a credential, as prescribed by OAuth2. In turn, the API can use Auth0 libraries (or any libraries capable to process JWT, the open standard Auth0 uses to encode access tokens) to verify the access tokens it receives from the calling application and issue a response with the desired data.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+The focus of this guide is to show you how to configure the SDK to call APIs that require an access token. Instead of creating a demo API to test the client-server connection, you'll use the [Auth0 Management API](https://auth0.com/docs/api/management/v2), which comes bundled with your Auth0 tenant. However, you can adapt this guide to work with any API that you are securing with Auth0.
+
+:::note
+If you followed the [previous section where you added user log in to Angular](/quickstart/spa/angular-next#add-login-to-your-application), make sure that you log out of your application as you'll need a new access token to call APIs.
+:::
+
+## Configure Your App to Make Authorized API Calls
+
+To facilitate calling APIs, the Angular SDK exports an [`HttpInterceptor`](https://angular.io/api/common/http/HttpInterceptor) that automatically adds an `Authorization` header with the access token to outgoing requests. However, only requests that are specified through configuration will have this header attached, preventing unintended leakage of access tokens to unexpected URLs.
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: i suggest (emphasis just here, not in the doc)

> the _appropriate_ access token

so that we don;t imply that there's only one

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+Once the above steps have been completed, any calls to APIs through Angular's `HttpClient` class to configured URLs will have an `Authorization` header attached.
+
+The following section demonstrates how to configure the application module to call APIs.
+
+### Configure the Auth Module
+
+The `AuthModule` setup is similar to the one discussed in the [Configure the `Auth0Provider` component](/quickstart/spa/angular-next##register-and-configure-the-authentication-module) section: you import the `AuthModule` type from the SDK into your Angular application module and call the static `forRoot` method, passing the `domain` and `clientId` properties. The values of these two props come from the ["Settings" values](https://auth0.com/docs/quickstart/spa/angular-next#configure-auth0) of the single-page application you've registered with Auth0.
+
+However, your Angular application needs to pass an access token when it calls a target API to access private resources. You can [request an access token](https://auth0.com/docs/tokens/guides/get-access-tokens) in a format that the API can verify by passing the `audience` and `scope` props to `AuthModule.forRoot`.
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: you can request an access token _at user authentication time_

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+In addition, the configuration should specify which requests should have an `Authorization` header added to them with the access token.
+
+The following example shows how to set up your Angular application module, with all the necessary imported types and the configuration to call the management API:
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: Does this overrides stuff that the reader might have written in the former chapter? Is there an alternative path in whihc this code is added to a clean starting point? In both cases, it seems the instructions should be more specific

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    }),
+  ],
+  providers: [
+    // Register the `AuthHttpInterceptor` to automatically attach the 'Authorization' header
+    { provide: HTTP_INTERCEPTORS, useClass: AuthHttpInterceptor, multi: true },
+  ],
+  bootstrap: [AppComponent],
+})
+```
+
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md:45]: if there's any way of referencing the section of the preceding chapter where the other config essential entries are explained, would be good to do so- or at least to mention that those explanations are there, and here we focus only on the new, API related stuff

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+Auth0 uses the value of the `audience` prop to determine which resource server (API) the user is authorizing your Angular application to access.
+
+Notice the `httpInterceptor` configuration, which specifies:
+
+* Any request where the URL starts with `https://${account.namespace}/api/v2/` should have an access token attached
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: longer term comment.
hmm, `account.namespace` looks odd... I am sure you are using this to be consistent with what we have done elsewhere: given that in other SDKs we use different terms (domain? tenant?) it might be worth to do a wide sweep to track what we use across the board, and begin a consolidation effort

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+:::note
+In the case of the Auth0 Management API, the audience is `https://${account.namespace}/api/v2/`. In the case of your APIs, you create an _Identifier_ value that serves as the _Audience_ value whenever you [set up an API](https://auth0.com/docs/getting-started/set-up-api) with Auth0.
+:::
+
+The actions that your Angular application can perform on the API depend on the [scopes](https://auth0.com/docs/scopes/current) that your access token contains, which you define as the value of `scope`. Your Angular application will request authorization from the user to access the requested scopes, and the user will approve or deny the request.
+
+:::note
+In the case of the Auth0 Management API, the `read:current_user` and `update:current_user_metadata` scopes let you get an access token that can retrieve user details and update the user's information. In the case of your APIs, you'll define custom [API scopes](https://auth0.com/docs/scopes/current/api-scopes) to implement access control, and you'll identify them in the calls that your client applications make to that API.
+:::
+
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md:98]: All the info above is useful, but I don;t know if it belongs to a quickstart... still feel a lot like docs to me. We need to have that info somewhere, but having it here seems to make the text longer.
Making the working assumption that we keep the current "doc" approach: this would be a good place to explain that all the settings described affect the AT you get at sign in time; if they want a different token for another API, or new scopes, they need to do something else (and hint at what)

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+## Make an API Call
+
+With your app module configured with the HTTP interceptor from the Angular SDK, calls you make to the Auth0 Management API will have a valid access token specified in the `Authorization` header. Let's use this as an example for showing user metadata.
```
Comment (vibronet) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: we can't know if it's valid, the API will decide that :) we can only know whether it's expired or not. We should briefly mention that we also refresh tokens automatically before use, when necessary, otherwise some developers will try to do it themselves and dig around for the RT

File: articles/quickstart/spa/angular-next/01-login.md [hunk 10]
```diff
@@ -107,17 +107,16 @@ import { AuthService } from '@auth0/auth0-angular';
...
-      *ngIf="(auth.isAuthenticated$ | async) === false"
-      (click)="auth.loginWithRedirect()">
-        Log in
-    </button>
-
-    <button 
-      *ngIf="auth.isAuthenticated$ | async"
-      (click)="auth.logout({ returnTo: 'http://localhost:4200' })">
+    <ng-container *ngIf="auth.isAuthenticated$ | async; else loggedOut">
+      <button (click)="auth.logout({ returnTo: 'http://localhost:4200' })">
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/01-login.md]: will the user remember this URL should target their app's base URL?

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+title: Call an API
+description: This tutorial demonstrates how to make API calls for protected resources on your server.
+budicon: 448
+topics:
+  - quickstarts
+  - spa
+  - angular
+  - api
+github:
+  path: Sample-01
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md:11]: haven't checked, but the same sample app as the first quickstart? 

```suggestion
  path: Sample-01
```

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 2 above)
Comment (lbalmaceda) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Separate note: if we leave "angular" aside, this paragraph will definitely match what we already have for other API articles. In fact, I think you've based yours in the [auth0-react](https://auth0.com/docs/quickstart/spa/react/02-calling-an-api) one. If we could rephrase this to make it platform agnostic (e.g. "your angular app" -> "your app"), we can reuse (docs' snippets) all the introductory text between quickstarts and focus on the code snippets or platform notes.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+* Register `AuthHttpInterceptor` in the `providers` section of your application module
+* Add configuration to specify which requests should have an `Authorization` header
+
+Once the above steps have been completed, any calls to APIs through Angular's `HttpClient` class to configured URLs will have an `Authorization` header attached.
+
+The following section demonstrates how to configure the application module to call APIs.
+
+### Configure the Auth Module
+
+The `AuthModule` setup is similar to the one discussed in the [Configure the `Auth0Provider` component](/quickstart/spa/angular-next##register-and-configure-the-authentication-module) section: you import the `AuthModule` type from the SDK into your Angular application module and call the static `forRoot` method, passing the `domain` and `clientId` properties. The values of these two props come from the ["Settings" values](https://auth0.com/docs/quickstart/spa/angular-next#configure-auth0) of the single-page application you've registered with Auth0.
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: I couldn't find this section above. Is this linking to another page? Also, is this double # on the link correct?
```suggestion
The `AuthModule` setup is similar to the one discussed in the [Configure the `Auth0Provider` component](/quickstart/spa/angular-next#register-and-configure-the-authentication-module) section: you import the `AuthModule` type from the SDK into your Angular application module and call the static `forRoot` method, passing the `domain` and `clientId` properties. The values of these two props come from the ["Settings" values](https://auth0.com/docs/quickstart/spa/angular-next#configure-auth0) of the single-page application you've registered with Auth0.
```

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 4 above)
Comment (lbalmaceda) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: nit: isn't "props" an angular term? this repeats across the doc

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 7 above)
Comment (lbalmaceda) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: `${account.namespace}` is the template string that the docs app will later process and replace with the "AUTH0_DOMAIN" or the actual app' domain if they are logged in.
Ref: https://github.com/auth0/docs/blob/master/CONTRIBUTING.md#user-specific-variables

Given that in this article Steve explains how to use the mgmt API, the audience must match the app's tenant.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+```
+
+Auth0 uses the value of the `audience` prop to determine which resource server (API) the user is authorizing your Angular application to access.
+
+Notice the `httpInterceptor` configuration, which specifies:
+
+* Any request where the URL starts with `https://${account.namespace}/api/v2/` should have an access token attached
+* If the URL is matched, the attached access token should have an audience value of `https://${account.namespace}/api/v2/` and the `read:current_user` scope;
+
+A more detailed run-down of the available configuration options for the HTTP interceptor is available in [the repository docs](https://github.com/auth0/auth0-angular#call-an-api).
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: ideally, we should keep everything in the docs site and not reference the github repo. We already have one for the spa sdk here https://auth0.com/docs/libraries. Consider in the future, adding one for the wrapper SDKs.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+
+Right now, the `metadata` property will always be empty. Populate the `ngOnInit` method to update this with code that fetches the user metadata from the Auth0 Management API:
+
+```js
+ngOnInit(): void {
+  this.auth.user$
+    .pipe(
+      concatMap((user) =>
+        this.http.get(
+          `https://angular-workshop.us.auth0.com/api/v2/users/<%= "${user.sub}" %>`
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: should this URL use the ${account.namespace} instead? 

also, given we're appending the -sub- value instead of using some utility method to construct the query, do we need to URL encode this? thinking of the pipe char

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 9 above)
Comment (lbalmaceda) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: I agree. Maybe we could use the "correct" access token instead.

**Review by lbalmaceda — CHANGES_REQUESTED** [review]
I don't have access to the preview link so I reviewed the .md file. In general looks good, I agree that we shouldn't need to explain in detail every auth0 product feature here. We could briefly mention them if absolutely necessary for context, and link them to external articles. Note that much of this content can be re-used across SPA articles so would be good to generify the paragraphs and just reference to those "snippet files". Also, there are 2 big code snippets (module setup and user profile component) that can be split into 2 or 3 more granular snippets. 

e.g. a first piece with the required option parameters, explaining how the default flow/scenario behaves. A second piece with some default overrides, for the second most common scenarios. And last, the third piece with optional parameters that go over the normal flow.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 2 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Thanks @vibronet. Yes this was mostly lifted from the React article as I believed that to be mostly correct, aside from the mention of Angular. You're right @lbalmaceda in that we could abstract this out and use in both places 👍

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 4 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Perhaps you meant _React_ here but yes - React has more or less coined "props" to mean something specific to React. I'll expand to "properties" throughout.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 8 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md:98]: Agreed, I'm happy to remove or change it. This is also present in the React quickstart, which I was using as a known good base.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 14 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Yes, good point on both counts. I've removed the absolute URL and used `encodeURI`. The request actually works either way, but encoding it is the right thing to do.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 13 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Yes, we are planning one. We do have one for React and are planning one for Angular. I'll leave the link until the document is live, but yes - let's link to library docs instead.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 12 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Yes it's linking to the first tutorial - I'll make the link more explicit.

I've applied the changes locally, nice spot on the double hash 👍

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 11 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md:11]: Yes - both articles link to the same sample, which demonstrates all the use cases together.

File: articles/quickstart/spa/angular-next/01-login.md (hunk 10 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/01-login.md]: I've modified this to get it generically rather than encode the URL, should cure that issue 👍

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+:::note
+If you followed the [previous section where you added user log in to Angular](/quickstart/spa/angular-next#add-login-to-your-application), make sure that you log out of your application as you'll need a new access token to call APIs.
+:::
+
+## Install the HTTP Interceptor
+
+To install and configure the HTTP interceptor, perform the following steps:
+
+* Import the `AuthHttpInterceptor` type from the Auth0 Angular SDK
+* Import `HttpClientModule` and `HTTP_INTERCEPTOR` from `@angular/common/http`
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: haven't checked locally but either this ends with S or the code snippets are wrong
```suggestion
* Import `HttpClientModule` and `HTTP_INTERCEPTORS` from `@angular/common/http`
```

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+* Register `AuthHttpInterceptor` in the `providers` section of your application module
+* Add configuration to specify which requests should have an `Authorization` header
+
+The following is an example of an Angular module (based upon the default implementation when you create a new app using `ng new`) that supports `AuthHttpInterceptor`, configured to call the Auth0 Management API with the ability to read the current user profile:
+
+```javascript
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: this is a really big chunk of code. I understand we want to give a lot of context around the lines that we're changing, but consider splitting this snippet into 2 or 3 smaller snippets that go between the steps that you are enumerating above. There's a good visual example of what I mean in the official angular docs https://angular.io/guide/http#provide-the-interceptor

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+        allowedList: [
+          {
+            // Match any request that starts 'https://${account.namespace}/api/v2/' (note the asterisk)
+            uri: 'https://${account.namespace}/api/v2/*'
+            tokenOptions: {
+              // The attached token should target this audience
+              audience: 'https://${account.namespace}/api/v2/',
+
+              // The attached token should have these scopes
+              scope: 'read:current_user'
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: I understand the base SDK is already enforcing the `openid` scope. The question is, do we want to explicitly request it for giving more visibility that the wrapper is OIDC conformant? Or at least add a comment to imply that these other set of scopes are also being included.

example (have not checked the default scopes)
```js
scope: 'read:current_user' // "openid profile email" are included internally
```

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+// Import the HttpClient for making API requests
+import { HttpClient } from '@angular/common/http';
+
+// Import AuthService from the Auth0 Angular SDK to get access to the user
+import { AuthService } from '@auth0/auth0-angular';
+
+@Component({
+  selector: 'app-metadata',
+  template: `<div *ngIf="auth.user$ | async as user">
```
Comment (lbalmaceda) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md]: the template is not using this `user` variable. I guess is better to either use it explicitly or await with ngIf on `metadata` instead.

**Review by lbalmaceda — CHANGES_REQUESTED** [review]

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 15 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Thanks, have applied this locally.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 17 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Good question. Here we're talking about scopes for the access token, so is it relevant that the extra default scopes are reference here?

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 18 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Good spot - I've gone with waiting on `metadata`.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 16 above)
Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Agreed, I like how the Angular docs do that. Let me break this snippet down in a similar way.

Comment (stevehobbsdev) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: I've broken it down now and pushed the changes - how does it look?

File: articles/quickstart/spa/angular-next/02-calling-an-api.md [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+          // The attached token should have these scopes
+          scope: 'read:current_user'
+        }
+      }
+    ]
+  }
+})
+```
+
+Please [refer to the docs](https://github.com/auth0/auth0-angular#configure-authhttpinterceptor-to-attach-access-tokens) for more information on the available options for the HTTP interceptor.
```
Comment (stevehobbsdev) [inline articles/quickstart/spa/angular-next/02-calling-an-api.md:97]: We can link to the library docs here when they are published. Right now, I have no idea what that link would be.

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 16 above)
Comment (lbalmaceda) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: Better, thanks!

File: articles/quickstart/spa/angular-next/02-calling-an-api.md (hunk 17 above)
Comment (lbalmaceda) [reply articles/quickstart/spa/angular-next/02-calling-an-api.md]: I asked out of curiosity. I don't have any preference. 👍

**Review by lbalmaceda — APPROVED** [review]
